File: 1]. DSA + CP/1]. DSA/1]. Data Structures/01]. Array/C++/_030)_Maximum_Circular_Subarray_Sum.cpp
#include<iostream>
using namespace std;

int normalMaxSum(int arr[], int n)
{
    int res = arr[0];
    int maxEnding = arr[0];
    for(int i=1; i<n; i++)
    {
        maxEnding = max(maxEnding + arr[i], arr[i]);
        res = max(res, maxEnding);
    }
    return res;
}

int maxCircularSubarraySum(int arr[], int n)
{
    int max_normal = normalMaxSum(arr,n);
    if(max_normal < 0)
        return 0;

    int arr_sum = 0;
    for(int i=0; i<n; i++)
    {
        arr_sum += arr[i];
        arr[i] = -arr[i];
    }
    int max_circular = arr_sum + normalMaxSum(arr, n);
    return max(max_normal, max_circular);
}

int main()
{
    int arr[] = {5, -2, 3, 4}, n=4;
    cout<<"Maximum Circular SubArray Sum = "<<maxCircularSubarraySum(arr,n)<<endl;
}

File: 1]. DSA + CP/1]. DSA/1]. Data Structures/01]. Array/Dart/_030)_Maximum_Circular_Subarray_Sum.dart
import 'dart:math';

int normalMaxSum(List<int> arr) {
  int res = arr[0];
  int maxEnding = arr[0];

  for (int i = 1; i < arr.length; i++) {
    maxEnding = max(maxEnding + arr[i], arr[i]);
    res = max(res, maxEnding);
  }
  return res;
}

int maxCircularSubArraySum(List<int> arr) {
  int maxNormal = normalMaxSum(arr);
  if (maxNormal < 0) return 0;

  int arrSum = 0;
  for (int i = 0; i < arr.length; i++) {
    arrSum += arr[i];
    arr[i] = -arr[i];
  }

  int maxCircular = arrSum + normalMaxSum(arr);
  return max(maxNormal, maxCircular);
}

void main() {
  var arr = [5, -2, 3, 4];
  print('Maximum circular sub-array sum is ${maxCircularSubArraySum(arr)}.');
}
